src/peft/tuners/tuners_utils.py

```diff
@@ -665,3 +688,43 @@ def check_adapters_to_merge(module: BaseTunerLayer, adapter_names: Optional[list
             warnings.warn("All adapters are already merged, nothing to do.")
 
     return adapter_names
+
+
+def clone_module(module: nn.Module, share_weights=False):
+    clone = copy.deepcopy(module)
+
+    def _share_weights(src: nn.Module, dst: nn.Module):
+        for name, param in src.named_parameters(recurse=False):
+            dst.register_parameter(name, param)
+
+    if share_weights:
+        for name, submodule in module.named_modules():
+            _share_weights(submodule, clone.get_submodule(name))
+
+    return clone
+
+
+def replicate_layers(model: nn.Module, layer_map: list[tuple[int, int]]):
+    """Replicate layers in a transfomer model with weight sharing.
+    
+    This function looks for a module list attribute at model[(.model)*].layers 
+    and replicates the layers in the module list according to the layer map.
+    For example the map `[[0, 4], [2, 5]]` will take the set of layers `[0, 1, 2, 3, 4]`
+    and replace them with a module list containing `[0, 1, 2, 3, 2, 3, 4]`.
+    """
+    while hasattr(model, 'model'):
+        model = model.model
+    if not hasattr(model, 'layers'):
```

Comment:
I wonder if we could put more effort into accommodating more architectures. E.g. for BERT it's called `"layer"`, for bloomz and Falcon, it just seems to be `"h"`, etc.

Reply:
I added support for bert and falcon. But have not figured out testing each of these.  Llama (and mistral) are much more common now so I am not sure how many we want to try and add. I know I should also update the test to cover these. Its a bit of work since I need to figure out dummy models for each. Pushing an update but I will look into the tests.